sycl/doc/CompilerAndRuntimeDesign.md

```diff
@@ -150,7 +177,70 @@ defines:
 
 - target triple and a native tool chain for each target (including "virtual"
 targets like SPIR-V).
-- SYCL offload action based on generic offload action
+- SYCL offload action based on generic offload action.
+
+SYCL compilation pipeline has a peculiarity compared to other compilation
+scenarios - some of the actions in the pipeline may output multiple "clusters"
+of files, other may input those. For example, each device binary maybe
+accompanied by a symbol table and a specialization constant map - additional
+information used by the SYCL runtime library - and it needs to be stored into
+the device binary descriptor by the offload wrapper tool. With device code
+splitting feature enabled, there can be multiple such sets (clusters) of files -
+one per each separate device binary.
+
+Clang driver's design can't handily model that, namely:
+1. Multiple inputs/outputs in the action graph.
+1. "Clusters" of inputs/outputs, when subsets of inputs/outputs are
+associated and describe different aspects of the same data.
+
+To support this, SYCL introduces the `file-table-tform` tool. This tool can
+transform file tables following commands passed as input arguments. Each row
+in the table represents a file cluster, each column - a type of data associated
+with a cluster. The tool can replace and extract columns. For example, the
+`sycl-post-link` tool can output two file clusters and the following file
+table referencing all the files in the clusters:
+```
+  [Code|Symbols|Properties]
+  a_0.bc|a_0.sym|a_0.props
+  a_1.bc|a_1.sym|a_1.props
+```
+
+When participating in the action graph this tool inputs a file table
+(`TY_Tempfiletable` clang input type) and/or a file list (`TY_Tempfilelist`),
+performs requested transformations and outputs a file table or list. From the
+clang design standpoint there is still single input and output, even though in
+reality there are multiple.
+
+For example, depending on compilation options, files from the "Code" column
+above may need to undergo AOT compilation after the device code splitting step,
+performed as a part of the code transformation sequence done by the
+`sycl-post-link` tool. The driver will then:
+- Use the `file-table-tform` to extract the code files and produce a file
+  list:
+```
+a_0.bc
+a_1.bc
+```
+- Pass this file list to the `llvm-for-each` tool along with AOT compilation
+command to invoke it on every file in the list. This will result in another
+file list
+```
+a_0.bin
+a_1.bin
+```
+- Then `file-table-tform` is invoked again to replace `.bc` with `.bin` in
+the filetable to get a new filetable:
+```
+  [Code|Symbols|Properties]
+  a_0.bin|a_0.sym|a_0.props
+  a_1.bin|a_1.sym|a_1.props
+```
+- Finally, this filetable is passed to the `clang-offfload-wrapper` tool to
+construct a wrapper object which embeds all those files.
+
+Note that the graph does not change with growing number of clusters, neither
+it changes when more files are added to each cluster (e.g. a "manifest" file).
```

Comment:
Please reword, are you trying to say that the graph also doesn't change when more files are added to each cluster?

Reply:
ok, fixed in latest push.